core/domain/user_services.py

```diff
@@ -2287,3 +2291,263 @@ def get_dashboard_stats(user_id):
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    if exploration_user_model is None:
+        exploration_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exploration_id)
+    exploration_user_model.most_recently_reached_checkpoint_exp_version = None
+    exploration_user_model.most_recently_reached_checkpoint_state_name = None
+    exploration_user_model.update_timestamps()
+    exploration_user_model.put()
+
+
+def synchronize_learner_checkpoint_progress_with_current_exp_version(
+    user_id,
+    exploration_id):
+    """Synchronizes the most recently reached checkpoint and the furthest
+    reached checkpoint with the latest exploration.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+
+    Returns:
+        ExplorationUserData. The domain object corresponding to the given user
+        and exploration.
+    """
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+
+    if exploration_user_model is not None:
+        latest_exploration = exp_fetchers.get_exploration_by_id(exploration_id)
+        most_recently_interacted_exploration = (
+            exp_fetchers.get_exploration_by_id(
+                exploration_id,
+                True,
+                (
+                    exploration_user_model
+                        .most_recently_reached_checkpoint_exp_version
+                )))
+        furthest_reached_exploration = (
+            exp_fetchers.get_exploration_by_id(
+                exploration_id,
+                True,
+                exploration_user_model.furthest_reached_checkpoint_exp_version
+            ))
+
+        most_recently_reached_checkpoint_in_current_exploration = _get_most_distant_reached_checkpoint_in_current_exploration( # pylint: disable=line-too-long
+            _get_checkpoints_in_order(
+                latest_exploration.init_state_name, latest_exploration.states),
+            _get_checkpoints_in_order(
+                most_recently_interacted_exploration.init_state_name,
+                most_recently_interacted_exploration.states),
+            exploration_user_model.most_recently_reached_checkpoint_state_name)
+
+        furthest_reached_checkpoint_in_current_exploration = _get_most_distant_reached_checkpoint_in_current_exploration( # pylint: disable=line-too-long
+            _get_checkpoints_in_order(
+                latest_exploration.init_state_name, latest_exploration.states),
+            _get_checkpoints_in_order(
+                furthest_reached_exploration.init_state_name,
+                furthest_reached_exploration.states),
+            exploration_user_model.furthest_reached_checkpoint_state_name)
+
+        # If the most recently reached checkpoint doesn't exist in current
+        # exploration.
+        if most_recently_reached_checkpoint_in_current_exploration != (
```

Comment:
condition seems inappropriate to me, we should update if there is difference in versions

Reply:
I have added the condition to compare exploration versions in the handler before calling out this method.